lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api/Types.hs

```diff
@@ -130,12 +157,30 @@ instance Accept Hex where
 instance FromJSON StakeApiResponse where
     parseJSON = genericParseJSON defaultOptions
 
+instance FromJSON ApiAccountState where
+    parseJSON = withObject "ApiAccountState" $ \v ->
+        ApiAccountState
+            <$> v .: "value"
+            <*> v .: "delegation"
+            <*> v .: "counter"
```

Comment:
What about using `Generic` and `genericParseJSON` instead here? This avoid encoding the structure twice (in the type itself, and in the JSON serializer/deserializer) and makes for code that is easier to maintain.

Reply:
@KtorZ wrote:
> What about using Generic and genericParseJSON instead here?

That would also be my preference (normally).

However, the names of the on-the-wire type are somewhat _cryptic_. For example, it's really not obvious that `counter` refers to a total count of the number of _transactions_.

Of course, we _could_ fix this by adding comments to the fields.

But every time we import one of these cryptic names into another module, the reader has to pay the mental cost of translating from the name to the meaning.

I'd argue that it's better to use _self-documenting names_ for the fields, and _pay the translation cost a single time_ in the JSON instances (which we will test).

There's also the issue that the on-the-wire format is subject to change. If the on-the-wire format changes and we use a generic JSON instance, then we have to change the names of the fields in all the modules that import them. But if we use a manually-defined JSON instance above, we can limit the change to just this instance.

Thoughts?

Reply:
@KtorZ I've revised the code to use generic JSON instances for everything, while retaining self-documenting field names on the exported type. See https://github.com/input-output-hk/cardano-wallet/pull/995/commits/06fe68e174c1fd0ec693303da51f39dd14e8af96

Reply:
> I'd argue that it's better to use self-documenting names for the fields, and pay the translation cost a single time in the JSON instances (which we will test).

I'd slightly disagree and would argue that it is best to stick with Jörmungandr terminology in the http-client for Jörmungandr. The very same way we try to keep the same terminology for the binary formats. This reduces the need for an extra mapping between our names and their names. 

Rather than the field, I'd recommend to make the *types* more expressive (as you did I believe :)) which is what matters the most in the end.